curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -0,0 +1,402 @@
+ */
+
+#include "curvefs/src/metaserver/s3compact_wq_impl.h"
+
+#include <algorithm>
+#include <deque>
+#include <list>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+using ::curve::common::Configuration;
+using ::curve::common::InitS3AdaptorOption;
+using ::curve::common::S3Adapter;
+using ::curve::common::S3AdapterOption;
+using ::curve::common::TaskThreadPool;
+
+namespace curvefs {
+namespace metaserver {
+
+void S3CompactWorkQueueImpl::Enqueue(const InodeKey& inodeKey) {
+    std::unique_lock<std::mutex> guard(mutex_);
+
+    // inodeKey already in working queue, just return
+    if (std::find(compactingInodes_.begin(), compactingInodes_.end(),
+                  inodeKey) != compactingInodes_.end()) {
+        return;
+    }
+
+    while (IsFullUnlock()) {
+        notFull_.wait(guard);
+    }
+    compactingInodes_.push_back(inodeKey);
+    auto task =
+        std::bind(&S3CompactWorkQueueImpl::CompactChunks, this, inodeKey);
+    queue_.push_back(std::move(task));
+    notEmpty_.notify_one();
+}
+
+std::function<void()> S3CompactWorkQueueImpl::Dequeue() {
+    std::unique_lock<std::mutex> guard(mutex_);
+    while (queue_.empty() && running_.load(std::memory_order_acquire)) {
+        notEmpty_.wait(guard);
+    }
+    std::function<void()> task;
+    if (!queue_.empty()) {
+        task = std::move(queue_.front());
+        queue_.pop_front();
+        compactingInodes_.pop_front();
+        notFull_.notify_one();
+    }
+
+    return task;
+}
+
+void S3CompactWorkQueueImpl::ThreadFunc() {
+    while (running_.load(std::memory_order_acquire)) {
+        Task task(Dequeue());
+        if (task) {
+            task();
+        }
+    }
+}
+
+std::vector<uint64_t> S3CompactWorkQueueImpl::getNeedCompact(
+    const ::google::protobuf::Map<uint64_t, S3ChunkInfoList>& s3chunkinfoMap) {
+    std::vector<uint64_t> needCompact;
+    int maxChunkPerIndex;
+    int fragmentThreshold;
+    LOG_IF(FATAL, !conf_->GetIntValue("s3compactwq.max_chunks_per_index",
+                                      &maxChunkPerIndex));
+    LOG_IF(FATAL, !conf_->GetIntValue("s3compactwq.fragment_threshold",
+                                      &fragmentThreshold));
+    for (const auto& item : s3chunkinfoMap) {
+        if (needCompact.size() >= maxChunkPerIndex) {
+            break;
+        }
+        if (item.second.s3chunks_size() > fragmentThreshold) {
+            needCompact.push_back(item.first);
+        }
+    }
+    return needCompact;
+}
+
+void S3CompactWorkQueueImpl::deleteObjs(const std::vector<std::string>& objs) {
+    for (const auto& obj : objs) {
+        const Aws::String aws_key(obj.c_str(), obj.size());
+        s3Adapter_->DeleteObject(aws_key);  // don't care success or not
+    }
+}
+
+std::list<struct S3CompactWorkQueueImpl::Node>
+S3CompactWorkQueueImpl::buildValidList(const S3ChunkInfoList& s3chunkinfolist,
+                                       uint64_t inodeLen) {
```

Comment:
add some code comments for this long, complex algorithm func.

Reply:
fixed.